source/common/decompressor/zlib_decompressor_impl.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "envoy/decompressor/decompressor.h"
+
+#include "zlib.h"
+
+namespace Envoy {
+namespace Decompressor {
+
+/**
+ * Implementation of decompressor's interface.
+ */
+class ZlibDecompressorImpl : public Decompressor {
+public:
+  ZlibDecompressorImpl();
+
+  /**
+   * Sets buffer size for feeding data to the decompressor routines.
+   * @param chunk amount of memory reserved for the compressor output.
+   */
+  ZlibDecompressorImpl(uint64_t chunk_size);
+
+  /**
+   * Init must be called in order to initialize the decompressor. Once decompressor is initialized,
+   * it cannot be initialized again. Init should run before decompressing any data.
+   * @param window_bits sets the size of the history buffer. It must be greater than or equal to
+   * the window_bits value provided when data was compressed (zlib manual).
+   */
+  void init(int8_t window_bits);
+
+  /**
+   * @return uint64_t CRC-32 if a gzip stream is being written or Adler-32 for other compression
+   * types.
```

Comment:
Here again can you elaborate the comment on when it is appropriate to call this? Also, I think it should say "read" instead of "written" for this to make sense on the decompressor side.